### PR TITLE
Astropy 3.1 and 3.2 support for {from,to}_pandas

### LINKED ISSUE #14: Move to_pandas/from_pandas improvements upstream
Rather than overloading to_pandas/from_pandas here, we should be able to implement support for handling ``Time`` columns in the core package directly.

### PR DESCRIPTION
**Pull request by nabobalis**
Not quite sure its the idea you had in mind but let's start here. 

I went ahead and reverted the commit for 3.1. But I added another Travis Job for astropy stable. I wasn't sure if I should reduce the travis jobs to just check for astropy stable and dev.

Closes #14

### CODE REVIEW

File: astropy_timeseries/sampled.py [hunk 1]
```diff
@@ -161,22 +171,38 @@ def from_pandas(self, df):
         # TODO: determine how user can specify time scale
         time = Time(df.index)
 
-        # Create table without the time column
-        table = Table.from_pandas(df)
+        if LooseVersion(astropy.__version__) > LooseVersion("3.1.1"):
```
Comment (astrofrog) [inline astropy_timeseries/sampled.py]: I think this should be <3.2 instead of >3.1.1 since the functionality is not going to be present in the 3.1 branch.

Comment (astrofrog) [reply astropy_timeseries/sampled.py]: Could you define ``ASTROPY_LT_32`` at the top of the file and then use that constant here?

File: astropy_timeseries/sampled.py [hunk 2]
```diff
@@ -161,22 +171,38 @@ def from_pandas(self, df):
...
-        Convert this time series to a :class:`~pandas.DataFrame` with a
-        :class:`~pandas.DatetimeIndex` index.
+        Convert this :class:`~astropy_timeseries.TimeSeries` to a
+        :class:`~pandas.DataFrame` with a :class:`~pandas.DatetimeIndex` index.
+
+        Parameters
+        ----------
+        index : None, bool, str
+            Specify DataFrame index mode
+            Only used if Astropy version >=``3.2``.
```
Comment (astrofrog) [inline astropy_timeseries/sampled.py]: Can you clarify this docstring? (the one in ``from_pandas`` is clearer, and it's not clear what passing a ``str`` would do here)

**Review by astrofrog — CHANGES_REQUESTED** [review]
This looks good, just a few comments below. In addition, could you split out the second commit into changes to the CI and changes to the rest of the code, then merge the two CI-related commits together and give it a message with a 🚫 in it? (this will make it easier to filter out those commits later once we are ready to merge into astropy)

File: astropy_timeseries/sampled.py (hunk 2 above)
Comment (nabobalis) [reply astropy_timeseries/sampled.py]: I had wondered if we should inherit the documentation string from Qtable.

File: astropy_timeseries/sampled.py [hunk 3]
```diff
@@ -161,22 +171,38 @@ def from_pandas(self, df):
...
+        dataframe : :class:`pandas.DataFrame`
+            A pandas :class:`pandas.DataFrame` instance
         """
 
         # Extract table without time column
         table = self[[x for x in self.colnames if x != 'time']]
 
         # First make a normal pandas dataframe
-        df = Table(table).to_pandas()
+        if LooseVersion(astropy.__version__) > LooseVersion("3.1.1"):
```
Comment (Cadair) [inline astropy_timeseries/sampled.py]: As an alternative to checking the version, you could inspect the function signature?

File: astropy_timeseries/sampled.py (hunk 2 above)
Comment (nabobalis) [reply astropy_timeseries/sampled.py]: I assume you would like some tests for the new index feature?

Comment (astrofrog) [reply astropy_timeseries/sampled.py]: Yes please, but this doesn't have to be too extensive.

File: astropy_timeseries/sampled.py (hunk 3 above)
Comment (astrofrog) [reply astropy_timeseries/sampled.py]: Let's just check the version with an ``ASTROPY_LT_32`` constant - once we merge this into astropy we can more easily search and remove then.

File: astropy_timeseries/tests/test_binned.py [hunk 4]
```diff
@@ -3,17 +3,11 @@
...
 
 from ..binned import BinnedTimeSeries
 
 
-INPUT_TIME = Time(['2016-03-22T12:30:31',
-                   '2015-01-21T12:30:32',
-                   '2016-03-22T12:30:40'])
-PLAIN_TABLE = Table([[1, 2, 11], [3, 4, 1], [1, 1, 1]], names=['a', 'b', 'c'])
-
-
```
Comment (nabobalis) [inline astropy_timeseries/tests/test_binned.py:16]: These were not used, would you like a version of this for the `binned` like the `sampled` ones?

File: astropy_timeseries/binned.py [hunk 5]
```diff
@@ -111,6 +116,7 @@ def __init__(self, data=None, time_bin_start=None, time_bin_end=None,
                     time_bin_end = times
                 time_bin_size = (time_bin_end - time_bin_start).sec * u.s
             elif time_bin_size is None:
+                # I am not sure this is triggerable.
```
Comment (astrofrog) [inline astropy_timeseries/binned.py]: @nabobalis - I can trigger it with:

```python
In [1]: from astropy_timeseries import BinnedTimeSeries

In [2]: BinnedTimeSeries(time_bin_start=['2016-03-22T12:30:31',
   ...:                                    '2016-03-22T12:30:34',
   ...:                                    '2016-03-22T12:30:37',
   ...:                                    '2016-03-22T12:30:39'])
   ...:                                    
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-2183cf3da47e> in <module>()
      2                                    '2016-03-22T12:30:34',
      3                                    '2016-03-22T12:30:37',
----> 4                                    '2016-03-22T12:30:39'])
      5 

~/miniconda3/envs/dev/lib/python3.7/site-packages/astropy_timeseries/binned.py in __init__(self, data, time_bin_start, time_bin_end, time_bin_size, n_bins, **kwargs)
     57 
     58         if time_bin_end is None and time_bin_size is None:
---> 59             raise TypeError("Either 'time_bin_size' or 'time_bin_end' should be specified")
     60 
     61         if not isinstance(time_bin_start, Time):

TypeError: Either 'time_bin_size' or 'time_bin_end' should be specified
```

File: astropy_timeseries/tests/test_binned.py (hunk 4 above)
Comment (astrofrog) [reply astropy_timeseries/tests/test_binned.py:16]: No I think it's fine as-is

File: astropy_timeseries/tests/test_sampled.py [hunk 6]
```diff
@@ -144,26 +149,36 @@ def test_fold():
...
     df1['a'] = [1, 2, 3]
     df1.set_index(pandas.DatetimeIndex(INPUT_TIME.datetime64), inplace=True)
 
     ts = TimeSeries.from_pandas(df1)
     assert_equal(ts.time.isot, INPUT_TIME.isot)
+    assert ts.colnames == ['time', 'a']
+
+    ts1 = TimeSeries.from_pandas(df1, index='a')
+    assert_equal(ts1.time.isot, INPUT_TIME.isot)
+    assert ts1.colnames == ['time', 'index', 'a']
```
Comment (astrofrog) [inline astropy_timeseries/tests/test_sampled.py]: So in this case, does it mean that ``a`` and ``time`` are both index columns? Maybe you can test ``ts1.indices`` to check that the expected columns are indices?

**Review by astrofrog — COMMENTED** [review]
Just a few minor comments, but I think this is almost good to go

File: astropy_timeseries/binned.py (hunk 5 above)
Comment (nabobalis) [reply astropy_timeseries/binned.py]: That seems to be the error from line 59, which I believe prevents this error line from being triggered.

File: astropy_timeseries/tests/test_sampled.py (hunk 6 above)
Comment (nabobalis) [reply astropy_timeseries/tests/test_sampled.py]: There was a mistake here, it should be a boolean here. 
I got confused when writing this test with to_pandas.

It just adds a copy of the time axis as a titled index column. 

Before:
```
ts
<TimeSeries length=3>
             time               a  
            object            int64
----------------------------- -----
2016-03-22T12:30:31.000000000     1
2015-01-21T12:30:32.000000000     2
2016-03-22T12:30:40.000000000     3
```

After

```
ts1
<TimeSeries length=3>
             time                      index            a  
            object                     object         int64
----------------------------- ----------------------- -----
2016-03-22T12:30:31.000000000 2016-03-22T12:30:31.000     1
2015-01-21T12:30:32.000000000 2015-01-21T12:30:32.000     2
2016-03-22T12:30:40.000000000 2016-03-22T12:30:40.000     3
```

If it is useful, I am not sure.

File: astropy_timeseries/binned.py (hunk 5 above)
Comment (astrofrog) [reply astropy_timeseries/binned.py]: @nabobalis - you are correct, can you remove this if branch?

**Review by astrofrog — CHANGES_REQUESTED** [review]
@nabobalis - sorry for the delay in reviewing this further. I think we shouldn't expose the ``index`` option in ``to_pandas``/``from_pandas`` and should always treat the time column as the index. Could you update that?

Also I noticed an issue in the astropy 3.2 case, where I end up with the time column being added as an index twice in ``from_pandas``:

```
In [4]: from astropy_timeseries import TimeSeries 
   ...: ts = TimeSeries(time=['2016-03-22T12:30:31','2016-03-22T12:30:34', '2016-03-22T12:30:37', '2016-03-22T12:30:39']) 
   ...: ts['flux'] = [1,2,3,4] 
   ...: df = ts.to_pandas()                                                                                                                            

In [5]: TimeSeries.from_pandas(df)                                                                                                                     
Out[5]: 
<TimeSeries length=4>
             time              flux
            object            int64
----------------------------- -----
2016-03-22T12:30:31.000000000     1
2016-03-22T12:30:34.000000000     2
2016-03-22T12:30:37.000000000     3
2016-03-22T12:30:39.000000000     4

In [6]: TimeSeries.from_pandas(df).indices                                                                                                             
Out[6]: 
[             time             rows
 ----------------------------- ----
 2016-03-22T12:30:31.000000000    0
 2016-03-22T12:30:34.000000000    1
 2016-03-22T12:30:37.000000000    2
 2016-03-22T12:30:39.000000000    3,              time             rows
 ----------------------------- ----
 2016-03-22T12:30:31.000000000    0
 2016-03-22T12:30:34.000000000    1
 2016-03-22T12:30:37.000000000    2
 2016-03-22T12:30:39.000000000    3]
```

Once this is fixed, we should be able to go ahead and merge this. Thanks!

File: astropy_timeseries/binned.py (hunk 5 above)
Comment (nabobalis) [reply astropy_timeseries/binned.py]: I removed just the elif.

**Comment by nabobalis** [comment]
@astrofrog, I think I have address those issues.

**Comment by astrofrog** [comment]
@nabobalis - thanks! One last request - since the CI in Travis already includes a check against astropy dev, and this PR should still work with Astropy 3.1, can you remove the CI changes from the commits?

**Comment by nabobalis** [comment]
I've removed all the CI changes.

**Review by astrofrog — APPROVED** [review]

**Comment by astrofrog** [comment]
Looks great, thanks!

**Pull request merged**